firmware/vectored_mtvec/main.rs

```diff
@@ -0,0 +1,98 @@
+#![no_std]
+#![no_main]
+
+use core::arch::{asm, global_asm};
+
+use miralis_abi::{failure, setup_binary, success};
+
+setup_binary!(main);
+
+fn main() -> ! {
+    // Set mtvec to vectored mode at base _raw_interrupt_trap_handler
+    // Then produce an interrupt (MTI) to fall inside the vector
+    // The redirection should end in success_trap_handler
+    unsafe {
+        let _raw_interrupt_trap_handler = _raw_interrupt_trap_handler as usize | 0b1;
+
+        asm!(
+            "csrw mtvec, {handler}",       // Setup trap handler
+            "csrw mideleg, 0",
+            "csrs mstatus, {mstatus_mie}", // Enable interrupts (MIE)
+            "csrs mie, {mtie}",            // Enable machine timer interrupt (MTIE)
+            handler = in(reg) _raw_interrupt_trap_handler,
+            mstatus_mie = in(reg) 0x8,
+            mtie = in(reg) 0x80,
+        );
+    }
+
+    // Setup a timer deadline in the past to trap directly
+    set_mtimecmp_future_value();
+
+    // The trap handler should exit, if we reach that point the handler did not do its job
+    log::error!("Firmware didn't trapped!");
+    failure()
```

Comment:
Hmmm technically there could be a delay because it is MMIO (Memory Mapped I/O) which is asynchronous, maybe we should spin a bit with a loop. For instance `10_000` iterations with a [`core::hint::spin_loop`](https://doc.rust-lang.org/stable/core/hint/fn.spin_loop.html#).

Reply:
I see, yes good idea